pkg/dyninst/gosym/symtab.go

```diff
@@ -627,6 +625,191 @@ func (lt *lineTable) locatePC(pc uint64, goFuncData []byte) []GoLocation {
+	offset := inlinedCallSize * int(fIdx)
+	if offset+parentPCOffset+4 > len(inlineTree) {
+		return 0, 0, 0, false
+	}
+	funcID = inlineTree[offset+funcIDOffset]
+	nameOff = binary.LittleEndian.Uint32(inlineTree[offset+functionNameOffOffset:])
+	parentPC = binary.LittleEndian.Uint32(inlineTree[offset+parentPCOffset:])
+	ok = true
+	return
+}
+
+type fileRange struct {
+	pcLo uint64
+	pcHi uint64
+	file string
+}
+
+func (lt *lineTable) fileRanges(f *GoFunction) ([]fileRange, error) {
+	offset, found := f.funcInfo.pcfile()
+	if !found {
+		return nil, fmt.Errorf("no file data for function %s", f.Name)
+	}
+	offset += uint32(lt.pcTab[0])
+	if int(offset) >= len(lt.data) {
+		return nil, fmt.Errorf("file data offset out of range for function %s", f.Name)
+	}
+
+	names := map[uint32]string{}
+	pc := f.Entry
+	var ranges []fileRange
+	for file := range iterPCValues(lt.data[offset:], pc, lt.quantum) {
+		fno := uint32(file.Val)
+		var name string
+		var ok bool
+		if name, ok = names[fno]; !ok {
+			name = lt.fileName(f.funcInfo, fno)
+			names[fno] = name
+		}
+		ranges = append(ranges, fileRange{
+			pcLo: pc,
+			pcHi: file.PC,
+			file: name,
+		})
+		pc = file.PC
+	}
+	return ranges, nil
+}
+
+type functionRange struct {
+	pcLo uint64
+	pcHi uint64
+	name string
+	file string
+}
+
+func (lt *lineTable) inlinedFunctionsMapping(f *GoFunction, goFuncData []byte) []functionRange {
+	inlineTree := f.funcInfo.funcData(funcdataInlTree, goFuncData)
+	if inlineTree == nil {
+		return nil
+	}
+
+	offset, found := f.funcInfo.pcDataStart(pcdataInlTreeIndex)
+	if !found {
+		return nil
+	}
+	offset += uint32(lt.pcTab[0])
+	if int(offset) >= len(lt.data) {
+		return nil
+	}
+
+	files, err := lt.fileRanges(f)
+	if err != nil {
+		return nil
+	}
+	fileIdx := 0
+
+	names := map[uint32]string{}
+	pc := f.Entry
+	var ranges []functionRange
+	for index := range iterPCValues(lt.data[offset:], pc, lt.quantum) {
+		// [pc, index.PC) maps to inline function data at index.Val.
+		var name string
+		if index.Val < 0 {
+			name = f.Name
+		} else {
+			funcID, nameOff, _, found := readInlTree(inlineTree, uint32(index.Val))
+			if !found || funcID == lt.wrapperFuncID {
+				pc = index.PC
+				continue
+			}
+			var ok bool
+			if name, ok = names[nameOff]; !ok {
+				name = lt.funcName(nameOff)
+				names[nameOff] = name
+			}
+		}
+		// Find the file that covers the beginning of the current PC range.
+		// It must be the same for the whole function range.
+		for fileIdx < len(files) && pc >= files[fileIdx].pcHi {
+			fileIdx++
+		}
+		file := unknownFile
+		if fileIdx < len(files) && files[fileIdx].pcLo <= pc {
+			file = files[fileIdx].file
+		}
+		ranges = append(ranges, functionRange{
+			pcLo: pc,
+			pcHi: index.PC,
+			name: name,
+			file: file,
+		})
+		pc = index.PC
+	}
+	return ranges
+}
+
+func (lt *lineTable) functionLines(pc uint64, goFuncData []byte) (map[string]FunctionLines, error) {
```

Comment:
could this function and others like it take in the `*GoSymbolTable` instead of the `goFuncData`? It's not very clear what `goFuncData` is.